### ISSUE TITLE
Transform and Custom BERT packer 

### ISSUE THREAD

**Issue by Daard** [tensorflow/tfx#5901]
I faced bug in the [TensorFlow Transform Repo](https://github.com/tensorflow/transform/) 

I want to create custom Transformer model and need my own WordPiece vocabulary, tokenizer, packer and preprocess layer. 

I used guides to create custom preprocessing layers: https://www.tensorflow.org/text/guide/subwords_tokenizer#export, https://www.tensorflow.org/text/guide/bert_preprocessing_guide#review

It works fine in the Jupyter Notebook, but I faced this Error during interactive run of Transform component:

```
ValueError                                Traceback (most recent call last)
~/miniconda3/envs/tfx_12/lib/python3.9/site-packages/apache_beam/runners/common.cpython-39-x86_64-linux-gnu.so in apache_beam.runners.common.DoFnRunner.process()

~/miniconda3/envs/tfx_12/lib/python3.9/site-packages/apache_beam/runners/common.cpython-39-x86_64-linux-gnu.so in apache_beam.runners.common.PerWindowInvoker.invoke_process()

~/miniconda3/envs/tfx_12/lib/python3.9/site-packages/apache_beam/runners/common.cpython-39-x86_64-linux-gnu.so in apache_beam.runners.common.PerWindowInvoker._invoke_process_per_window()

~/miniconda3/envs/tfx_12/lib/python3.9/site-packages/apache_beam/transforms/core.py in <lambda>(x, *args, **kwargs)
   1876   if fn_takes_side_inputs(fn):
-> 1877     wrapper = lambda x, *args, **kwargs: [fn(x, *args, **kwargs)]
   1878   else:

~/miniconda3/envs/tfx_12/lib/python3.9/site-packages/tensorflow_transform/beam/impl.py in _convert_to_record_batch(batch_dict, schema, converter, passthrough_keys, input_metadata)
    477 
--> 478   arrow_columns, arrow_schema = impl_helper.convert_to_arrow(
    479       schema, converter, batch_dict)

~/miniconda3/envs/tfx_12/lib/python3.9/site-packages/tensorflow_transform/impl_helper.py in convert_to_arrow(schema, converter, fetches)
    601       tensors[name] = tensor_or_value
--> 602   record_batch = converter.convert(tensors)
    603   arrow_schema = record_batch.schema

~/miniconda3/envs/tfx_12/lib/python3.9/site-packages/tfx_bsl/tfxio/tensor_to_arrow.py in convert(self, tensors)
    129 
--> 130     return pa.record_batch(arrays, schema=self._arrow_schema)
    131 

~/miniconda3/envs/tfx_12/lib/python3.9/site-packages/pyarrow/table.pxi in pyarrow.lib.record_batch()

~/miniconda3/envs/tfx_12/lib/python3.9/site-packages/pyarrow/table.pxi in pyarrow.lib.RecordBatch.from_arrays()

ValueError: Arrays were not all the same length: 1 vs 2
```

**System information**

- Code 

```
import re
import pathlib
from typing import Dict, Any, List

import tensorflow as tf
import tensorflow_text as tf_text

import tensorflow_hub as hub
import tensorflow as tf
import tensorflow_transform as tft

def cleanup_text(reserved_tokens, token_txt):
    # Drop the reserved tokens, except for "[UNK]".
    bad_tokens = [re.escape(tok) for tok in reserved_tokens if tok != "[UNK]"]
    bad_token_re = "|".join(bad_tokens)
    bad_cells = tf.strings.regex_full_match(token_txt, bad_token_re)
    result = tf.ragged.boolean_mask(token_txt, ~bad_cells)
    # Join them into strings.
    result = tf.strings.reduce_join(result, separator=' ', axis=-1)
    return result


class CustomTokenizer(tf.Module):
    def __init__(self, reserved_tokens, vocab_path):
        self.tokenizer = tf_text.BertTokenizer(vocab_path, lower_case=True)
        self._reserved_tokens = reserved_tokens

    def tokenize(self, strings):
        enc = self.tokenizer.tokenize(strings)
        enc = enc.merge_dims(-2,-1)
        return enc

    # @tf.function
    def detokenize(self, tokenized):
        words = self.tokenizer.detokenize(tokenized)
        return cleanup_text(self._reserved_tokens, words)


class BertPackInputs(tf.keras.layers.Layer):

    def __init__(self,
                 seq_length,
                 start_of_sequence_id=None,
                 end_of_segment_id=None,
                 **kwargs):
        super().__init__(**kwargs)
        self.seq_length = seq_length
        self.start_of_sequence_id = start_of_sequence_id
        self.end_of_segment_id = end_of_segment_id

    def get_config(self) -> Dict[str, Any]:
        config = super().get_config()
        config["seq_length"] = self.seq_length
        config["start_of_sequence_id"] = self.start_of_sequence_id
        config["end_of_segment_id"] = self.end_of_segment_id
        return config

    def call(self, inputs: List[tf.RaggedTensor], *args, **kwargs):
        trimmed_segments = tf_text.RoundRobinTrimmer(max_seq_length=self.seq_length - 2).trim(inputs)
        segments_combined, segment_ids = tf_text.combine_segments(
            trimmed_segments,
            start_of_sequence_id=self.start_of_sequence_id,
            end_of_segment_id=self.end_of_segment_id)
        input_word_ids, input_mask = tf_text.pad_model_inputs(segments_combined, max_seq_length=self.seq_length)
        input_type_ids, _ = tf_text.pad_model_inputs(
            segment_ids, max_seq_length=self.seq_length)
        return dict(input_word_ids=input_word_ids,
                    input_mask=input_mask,
                    input_type_ids=input_type_ids)


class CustomPreprocess(tf.keras.layers.Layer):

    def __init__(self, reserved_tokens, vocab_path, max_len, start_id, end_id):
        super(CustomPreprocess, self).__init__()
        self.tokenizer = CustomTokenizer(reserved_tokens, vocab_path)
        self.packer = BertPackInputs(max_len, start_id, end_id)

    def call(self, inputs, *args, **kwargs):
        tokenized = self.tokenizer.tokenize(inputs)
        return self.packer([tokenized])

# tf.config.run_functions_eagerly(True)

def preprocessing_fn(inputs):
    
    reserved_tokens = ['[PAD]', '[UNK]', '[START]', '[END]', '[MASK]', '[RANDOM]', '[CLS]', '[SEP]']
    custom_prerocess= CustomPreprocess(reserved_tokens,
                                       './datasets3/page_vocab.txt', 
                                       128, 6, 3)
    
    def bert_preprocess(texts, name):
        preps = custom_prerocess([tf.squeeze(texts, axis=1)])
        return {f'{name}_word_ids': tf.cast(preps["input_word_ids"], tf.int64),
            f'{name}_mask': tf.cast(preps["input_mask"], tf.int64),
            f'{name}_type_ids': tf.cast(preps["input_type_ids"], tf.int64)}
    
    params = {}
    for suf in ['_x', '_y']:
        params['page_id' + suf] = inputs['page_id' + suf]
        
        for column in ['price', 'room_count']:
            params[column + suf] = tft.scale_to_z_score(inputs[column + suf])
        params['district' + suf] = tft.compute_and_apply_vocabulary(inputs['district' + suf],
                                                                    num_oov_buckets=1,
                                                                    vocab_filename='district_vocab' + suf)
        params.update(bert_preprocess(inputs['text' + suf], suf[1]))
    
    return params
```

- Environment in which the code is executed (e.g., Local Linux, Google Cloud, etc):
- TensorFlow version: 2.11
- TFX Version: 1.12
- Python version: 3.9
- Python dependencies (from `pip freeze` output):
```
tensorboard==2.11.0
tensorboard-data-server==0.6.1
tensorboard-plugin-wit==1.8.1
tensorflow==2.11.0
tensorflow-addons==0.19.0
tensorflow-data-validation==1.12.0
tensorflow-datasets==4.8.0
tensorflow-estimator==2.11.0
tensorflow-hub==0.12.0
tensorflow-io-gcs-filesystem==0.29.0
tensorflow-metadata==1.12.0
tensorflow-model-analysis==0.43.0
tensorflow-model-optimization==0.7.3
tensorflow-recommenders==0.7.2
tensorflow-serving-api==2.11.0
tensorflow-text==2.11.0
tensorflow-transform==1.12.0
```

How should I solve it?

I tried tfhub pretrained preprocess and tokenizer and it works fine in similiar function: 

```
class ModifiedBertPreprocess(tf.keras.layers.Layer):
    def __init__(self, max_len):
        super(ModifiedBertPreprocess, self).__init__()
        
        tfhub_handle_preprocess = 'https://tfhub.dev/tensorflow/bert_multi_cased_preprocess/3'
        preprocessor = hub.load(tfhub_handle_preprocess)
        
        self.tokenizer = hub.KerasLayer(preprocessor.tokenize, name="tokenizer")
        self.prep_layer = hub.KerasLayer(
                             preprocessor.bert_pack_inputs,
                             arguments={"seq_length":max_len})
        
    def call(self, inputs, training):
        tokenized = [self.tokenizer(seq) for seq in inputs]
        return self.prep_layer(tokenized)
```

**Comment by Daard** [comment]
Actually, you did not run my code. Cause you did not add %%writefile {_movies_transform_module_file} to the cell with my code, thus you reused vanilla prerocessing from the guide.

I built custom vocabulary for my dataset. You may use this code for building for movie dataset. But replace vocabulary path inside CustomPreprocess - **./datasets3/page_vocab.txt**. Replace **text_ds** for **movie ds** and select right feature inside lambda

```
bert_tokenizer_params=dict(lower_case=False)
reserved_tokens=["[PAD]", "[UNK]", "[START]", "[END]", "[MASK]", "[RANDOM]", "[CLS]", "[SEP]"]

bert_vocab_args = dict(
    # The target vocabulary size
    vocab_size = 8000,
    # Reserved tokens that must be included in the vocabulary
    reserved_tokens=reserved_tokens,
    # Arguments for `text.BertTokenizer`
    bert_tokenizer_params=bert_tokenizer_params,
    # Arguments for `wordpiece_vocab.wordpiece_tokenizer_learner_lib.learn`
    learn_params={},
)

page_vocab = bert_vocab.bert_vocab_from_dataset(
    **text_ds**.map(**lambda x: x['text']**).batch(1000).prefetch(2),
    **bert_vocab_args
)

def write_vocab_file(filepath, vocab):
    with open(filepath, 'w') as f:
        for token in vocab:
            print(token, file=f)
            
write_vocab_file('**./datasets3/page_vocab.txt'**, page_vocab)
```

**Comment by iindyk** [comment]
It seems that your preprocessing_fn returns a dictionary of tensors of different lengths. First dimension of all outputs is treated as batch dimension, so it must be the same

**Comment by Daard** [comment]
I have checked this code again. Bert preprocessing dict returns three dictionaries: tokens, mask and token types. The last one is always zeros. For some reason TFX removes extra dimensions of tensors with only zeros. 

For example: [[0, 0, 0], [0, 0, 0]] becomes [0, 0]. 

If I drop such tensors it will remove the problem. 

I have investigated this by subsampling my dataset and checking the results of the Transformer gen. 
Is there more efficient method?
